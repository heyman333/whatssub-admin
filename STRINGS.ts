const STRINGS = {
  en: { // English
    BUTTON: 'Button',
    SIGNUP: 'Sign Up',
    LOGIN: 'Login',
    LOGOUT: 'Logout',
    GOOGLE_LOGIN: 'Login with Google',
    FACEBOOK_LOGIN: 'Login with Facebook',
    EMAIL: 'Email',
    EMAIL_HINT: 'Write email address.',
    EMAIL_ERROR: 'Wrong email address',
    PASSWORD: 'Password',
    PASSWORD_HINT: 'Write password.',
    COMPLETE: 'Done',
    GOTO_Temp: 'No Page',
    BACK: 'Back',
    NAVIGATE: 'Navigate',
    CHANGE_THEME: 'Change theme',
    SIGN_IN_SCREEN: 'Sign in screen',
    MAIN_SCREEN: 'Main screen',
    UPDATE_BUTTON: 'Update',
    DELETE_BUTTON: 'Delete',
    SERVICE_LIST: 'Service List',
    SERVICE_ADD_BUTTON: 'Add Service',
    PRODUCT_LIST: 'Product List',
    PRODUCT_ADD_BUTTON: 'Add Product',
    SAVE: 'Save',
    CANCEL: 'Cancel',
    ICON: 'Icon',
    SERVICE_NAME: 'Service Name',
    SERVICE_NAME_KOR: 'Service Name(KOR)',
    WEB_PAGE: 'Web Page',
    DESCRIPTION: 'Description',
    PNG_GIF_ONLY: 'you can upload png and gif files only',
    INPUT_REQUIRED: 'is required',
    INPUT_MAX_VALIDATION: 'you exceeded max length of',
  },
  ko: { // Korean
    BUTTON: '버튼',
    SIGNUP: '회원가입',
    LOGIN: '로그인',
    LOGOUT: '로그아웃',
    GOOGLE_LOGIN: '구글 로그인',
    FACEBOOK_LOGIN: '페이스북 로그인',
    EMAIL: '이메일',
    EMAIL_ERROR: '이메일이 잘못 입력되었습니다.',
    EMAIL_HINT: '이메일 주소를 입력해주세요.',
    PASSWORD: '비밀번호',
    PASSWORD_HINT: '비밀번호를 입력해주세요.',
    COMPLETE: '완료',
    GOTO_Temp: '없는 페이지',
    BACK: '뒤로',
    NAVIGATE: '이동하기',
    CHANGE_THEME: '테마 변경',
    SIGN_IN_SCREEN: '로그인 화면',
    MAIN_SCREEN: '메인 화면',
    UPDATE_BUTTON: '수정',
    DELETE_BUTTON: '삭제',
    SERVICE_LIST: '서비스 목록',
    SERVICE_ADD_BUTTON: '서비스 추가',
    PRODUCT_LIST: '상품 목록',
    PRODUCT_ADD_BUTTON: '상품 추가',
    SAVE: '저장',
    CANCEL: '취소',
    ICON: '아이콘',
    SERVICE_NAME: '서비스명',
    SERVICE_NAME_KOR: '서비스명(한글)',
    WEB_PAGE: '웹페이지',
    DESCRIPTION: '설명',
    PNG_GIF_ONLY: 'png파일과 gif파일만 업로드 가능합니다',
    INPUT_REQUIRED: '의 입력이 필요합니다',
    INPUT_MAX_VALIDATION: '해당 길이 이상으로 입력될 수 없습니다:',
  },
  // ja: { // Japanese
  // },
  // zh: { // Chinese
  // },
  // es: { // Spanish
  // },
  // fr: { // French
  // },
  // id: { // Indonesian
  // },
};

const DEFAULT_LANG = 'en';
let lang = DEFAULT_LANG;

if (navigator) {
  lang = navigator.language.substr(0, 2);
}

export const getString = (str: string) => {
  str = STRINGS[lang]
    ? STRINGS[lang][str]
    : STRINGS[DEFAULT_LANG][str];

  if (!str) {
    str = '...';
  }
  return str;
};

export default STRINGS;
