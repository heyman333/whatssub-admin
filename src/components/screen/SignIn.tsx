import React, { Component, useState } from 'react';
import styled from 'styled-components';
import { getString } from '../../../STRINGS';
import { Formik } from 'formik';

// SignIn Page-wide font-size. styles resize proportionally by changing it
const fontSize = `20px`;

const Container = styled.div`
  display: flex;
  height: 100vh;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: ${(props) => props.theme.background};
  color: ${(props) => props.theme.fontColor};
  padding: 50px;
  font-size: ${fontSize};
`;

const Form = styled.form`
  display: flex;
  flex-direction: column;
`;

const Title = styled.div`
  font-size: 2em;
  margin-bottom: 1em;
`;

const Label = styled.label`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  font-size: 1em;
  margin: 0.5em 0;
`;

const Input = styled.input`
  font-size: 1em;
  width: 15em;
  height: 2em;
  padding: 0.3em;
`;

const SubmitButton = styled.button`
  align-self: flex-end;
  padding: 0.3em 0.5em;
  font-size: 0.9em;
`;

const ErrorMessage = styled.div`
  color: red;
  font-size: 0.7em;
`;

interface IProps {
  history?: any;
  onSubmitSuccess?: () => void // test purpose
}

function SignIn(props: IProps) {
  return (
    <Container>
      <Title>Whatssub Admin</Title>

      <Formik
        initialValues={{ email: '', password: '' }}
        validate={(values) => {
          let errors: { email?: string, password?: string } = {};

          // email validation
          if (!values.email) {
            errors.email = getString('EMAIL_HINT');
          } else if (
            !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)
          ) {
            errors.email = getString('EMAIL_ERROR');
          }

          // password validation
          if (!values.password) {
            errors.password = getString('PASSWORD_HINT');
          }
          return errors;
        }}
        onSubmit={(values, { setSubmitting }) => {
          if (props.onSubmitSuccess) {
            props.onSubmitSuccess();
          }

          setTimeout(() => {
            setSubmitting(false);
          }, 400);
        }}
      >
        {({
          values,
          errors,
          touched,
          handleChange,
          handleBlur,
          handleSubmit,
          isSubmitting,
        }) => (
          <Form onSubmit={handleSubmit}>
            <Label >
                Email:
              <Input
                type="email"
                name="email"
                data-testid="email"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.email}
              />
            </Label>
            {errors.email && touched.email && <ErrorMessage data-testid='error'>{errors.email}</ErrorMessage>}
            <Label>
                Password:
              <Input
                type="password"
                name="password"
                data-testid="password"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.password}
              />
            </Label>
            {errors.password && touched.password && <ErrorMessage>{errors.password}</ErrorMessage>}
            <SubmitButton type="submit" data-testid="submit" disabled={isSubmitting}>
              {getString('LOGIN')}
            </SubmitButton>
          </Form>
        )}
      </Formik>
    </Container>
  );
}

export default SignIn;
