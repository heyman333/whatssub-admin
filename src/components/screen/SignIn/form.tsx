import React, { Component, useState } from 'react';
import styled from 'styled-components';
import { getString } from '../../../../STRINGS';

const Form = styled.form`
  display: flex;
  flex-direction: column;
`;

const Label = styled.label`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  font-size: 1em;
  margin: 0.5em 0;
`;

const Input = styled.input`
  font-size: 1em;
  width: 15em;
  height: 2em;
  padding: 0.3em;
`;

const SubmitButton = styled.button`
  align-self: flex-end;
  padding: 0.3em 0.5em;
  font-size: 0.9em;
`;

const ErrorMessage = styled.div`
  color: red;
  font-size: 0.7em;
`;

const SignInForm = ({
  values,
  errors,
  touched,
  handleChange,
  handleBlur,
  handleSubmit,
  isSubmitting,
}) => (
  <Form onSubmit={handleSubmit}>
    <Label >
      Email:
      <Input
        type="email"
        name="email"
        data-testid="email"
        onChange={handleChange}
        onBlur={handleBlur}
        value={values.email}
      />
    </Label>
    {errors.email && touched.email && <ErrorMessage data-testid='error'>{errors.email}</ErrorMessage>}
    <Label>
      Password:
      <Input
        type="password"
        name="password"
        data-testid="password"
        onChange={handleChange}
        onBlur={handleBlur}
        value={values.password}
      />
    </Label>
    {errors.password && touched.password && <ErrorMessage>{errors.password}</ErrorMessage>}
    <SubmitButton type="submit" data-testid="submit" disabled={isSubmitting}>
      {getString('LOGIN')}
    </SubmitButton>
  </Form>
);
export default SignInForm;
