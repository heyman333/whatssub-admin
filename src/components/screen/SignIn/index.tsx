import React, { Component, useState } from 'react';
import styled from 'styled-components';
import { getString } from '../../../../STRINGS';
import { Formik } from 'formik';
import SignInForm from './form';

// SignIn Page-wide font-size. styles resize proportionally by changing it
const fontSize = `20px`;

const Container = styled.div`
  display: flex;
  height: 100vh;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: ${(props) => props.theme.background};
  color: ${(props) => props.theme.fontColor};
  padding: 50px;
  font-size: ${fontSize};
`;

const Title = styled.div`
  font-size: 2em;
  margin-bottom: 1em;
`;

interface IProps {
  history?: any;
  onSubmitSuccess?: () => void // test purpose
}

function SignIn(props: IProps) {
  return (
    <Container>
      <Title>Whatssub Admin</Title>

      <Formik
        initialValues={{ email: '', password: '' }}
        validate={(values) => {
          let errors: { email?: string, password?: string } = {};

          // email validation
          if (!values.email) {
            errors.email = getString('EMAIL_HINT');
          } else if (
            !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)
          ) {
            errors.email = getString('EMAIL_ERROR');
          }

          // password validation
          if (!values.password) {
            errors.password = getString('PASSWORD_HINT');
          }
          return errors;
        }}
        onSubmit={(values, { setSubmitting }) => {
          if (props.onSubmitSuccess) {
            props.onSubmitSuccess();
          }

          setTimeout(() => {
            setSubmitting(false);
          }, 400);
        }}
        component={(props) => <SignInForm {...props} />}
      />
    </Container>
  );
}

export default SignIn;
