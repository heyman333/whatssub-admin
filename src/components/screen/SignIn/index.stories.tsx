import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { linkTo } from '@storybook/addon-links';
import SignIn from '.';

storiesOf('SignIn', module)
  .add('default', () => (
    <SignIn
      onSubmitSuccess={() => { console.log('Success'); }}
    />));
