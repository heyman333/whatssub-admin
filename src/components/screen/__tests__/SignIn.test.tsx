import React from 'react';
import * as renderer from 'react-test-renderer';

import SignIn from '../SignIn';
import Button from '../../shared/Button';
import { render, fireEvent, getByTestId, wait, queryByTestId } from '@testing-library/react';
import { getString } from '../../../../STRINGS';

const props = {
  onSubmitSuccess: jest.fn()
};

describe('[SignIn] render', () => {
  it('renders without crashing', () => {
    const rendered = renderer.create(<SignIn />).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

describe('[SignIn] Interaction', () => {
  const component = <SignIn {...props} />;
  let renderResult: any;

  beforeAll(() => {
    renderResult = render(component);
  });

  it('[onSubmit] of Formik should not be called when submit button is clicked with empty inputs', async () => {
    const emailInput = renderResult.getByTestId('email');
    expect(emailInput.value).toBe('');
    const passwordInput = renderResult.getByTestId('password');
    expect(passwordInput.value).toBe('');
    const submitButton = renderResult.getByTestId('submit');
    fireEvent.click(submitButton);

    //formik is called asynchronously
    await wait(() => {
      expect(props.onSubmitSuccess).toHaveBeenCalledTimes(0);
    })
  });

  it('[onSubmit] of Formik should be called when [submit] button is clicked with valid input values', async () => {
    const email = 'whatssub123@gmail.com';
    const password = '1234';

    const emailInput = renderResult.getByTestId('email');
    fireEvent.change(emailInput, { target: { value: email } })
    expect(emailInput.value).toBe(email);

    const passwordInput = renderResult.getByTestId('password');
    fireEvent.change(passwordInput, { target: { value: password } })
    expect(passwordInput.value).toBe(password);

    const submitButton = renderResult.getByTestId('submit');
    fireEvent.click(submitButton);

    //formik is called asynchronously
    await wait(() => {
      expect(props.onSubmitSuccess).toHaveBeenCalledTimes(1);
    })
  })

  it('has email [error] massage when email is not valid', async ()=> {
    const email = 'notVailidEmail';

    //error message dosen't exist at first
    expect(renderResult.queryByTestId('error')).toBeNull()

    const emailInput = renderResult.getByTestId('email');
    fireEvent.focus(emailInput);
    fireEvent.change(emailInput, { target: { value: email }});
    fireEvent.blur(emailInput);

    await wait(()=>{
      const errorMessage = renderResult.getByTestId('error');
      expect(errorMessage).toBeDefined();
      expect(errorMessage.textContent).toBe(getString('EMAIL_ERROR'))
    })
  })

});
