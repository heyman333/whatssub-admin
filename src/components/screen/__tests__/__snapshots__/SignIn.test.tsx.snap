// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SignIn] render renders without crashing 1`] = `
<div
  className="sc-EHOje esHauf"
>
  <div
    className="sc-bZQynM cndpHV"
  >
    Whatssub Admin
  </div>
  <form
    className="sc-bdVaJa bPGLpq"
    onSubmit={[Function]}
  >
    <label
      className="sc-bwzfXH jYEUDA"
    >
      Email:
      <input
        className="sc-htpNat cymWMm"
        data-testid="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        type="email"
        value=""
      />
    </label>
    <label
      className="sc-bwzfXH jYEUDA"
    >
      Password:
      <input
        className="sc-htpNat cymWMm"
        data-testid="password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        type="password"
        value=""
      />
    </label>
    <button
      className="sc-bxivhb hFrwfb"
      data-testid="submit"
      disabled={false}
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
